text/3371-cargo-target-dir-templates.md

```diff
@@ -0,0 +1,424 @@
+
+[source]: https://github.com/rust-lang/cargo/issues/11156#issuecomment-1285951209
+
+## Remapping
+
+[rust-lang/cargo#11156](https://github.com/rust-lang/cargo/issues/11156) was originally about remapping the target directory, not about having a central one but reading the issue, there seems to be no needs for more than the simple redefinition of the target directory proposed in this document. In the future, if `CARGO_TARGET_DIR_REMAP` is introduced, it could be used to be the prefix to the target directory like so:
+
+- Set `CARGO_TARGET_DIR_REMAP=/home/user/projects=/tmp/cargo-build`
+- Compile the crate under `/home/user/projects/foo/` **without** `CARGO_TARGET_DIR` set
+- The resulting target directory will be at `/tmp/cargo-build/foo/target`
+
+By making the priority order `CARGO_TARGET_DIR` > `CARGO_TARGET_DIR_REMAP` (when both are absolute paths) we would keep backward compatibility. Or we could disallow having the two set at once, so that they're alternatives and not ordered.
+
+When `CARGO_TARGET_DIR` is relative, the result could be `/tmp/cargo-build/foo/$CARGO_TARGET_DIR`.
+
+Overall, I feel remapping is much harder to implement well and can be added later without interfering with templates in `CARGO_TARGET_DIR` (and without this RFC interfering with remapping), though the design space is probably bigger than the one for this RFC.
+
+## Advantages of using a hash over regular subdirectories
+
+It's possible to achieve most of the same system proposed here by setting a value like `CARGO_TARGET_DIR="/base/dir/{manifest-path-dirs}"`, where a manifest in `/tmp/test1/test2/Cargo.toml` would resolve the build directory to `/base/dir/tmp/test1/test2/`, but most is not *all* of them:
+
+1. Hashes have a fixed length while `manifest-path-dirs` is dependent on the context, making it a hazard for cross-platform compatibility. Say on Windows the target-dir is rooted in a user tmp dir and the manifest path is inside of the user documents. Especially combined with corporate policies on names, those base paths alone can take up a good amount of the character budget without getting into project names, etc.
+2. Encourage interactions through `cargo-metadata`: by using paths computed through cargo and not easily derivable from the file tree, future tools will be incentivized to work through `cargo-metadata` to find the target directory, widening adoption and making it easier for the cargo team to ensure nothing breaks in subsequent cargo updates
+3. Avoid introducing a strong dependency on only the path: by using a hash, cargo can add element to it to help differentiate builds: for example we could use more parameters in the hash, see the relevant section in Future Possibilites
+
+## Alternatives to the transition period
+
+While unlikely, the transition period may break workflows by introducing errors for previously valid target directories.
+Several alternatives exist for this:
```

Comment:
It would be worth talking about what happens if the config option is turned on and off, or if older and newer Cargo versions are invoked with e.g. `cargo +version build`. And to consider the behavior both with and without forward links. For example, let's say Cargo version 1.500 supports templating:

* Fresh build with Cargo 1.500
* Fresh build with 1.500 and then with Rust 1.499
* Fresh build with 1.499 and then with Rust 1.500
* `cargo +1.499 clean`

Old Rust versions will be around for a while, and typical user behavior would be to set it in the global env or .cargo/config, and then forget about it.

For comparison, targo needs to be invoked once to create the symlink, and then most cargo commands work fine as usual regardless of version.

Reply:
> It would be worth being explicit

On the contrary, I don't think it's worth it: that would be better determined by experimentation I think, and once we have something of a consensus by testers we can finalize and write a definite table for it.

There are effectively two options:

- Templating
- Forwarding

The second does not matter if the first is inactive.

If templating is active it means `CARGO_TARGET_DIR` is set in some way, in which case the behaviour is known: `cargo` creates a directory as directed, so `CARGO_TARGET_DIR=/tmp/{target}` would create exactly `/tmp/{target}`, as it does today already.

I have a few rules I would like to see because they're also how `cargo` operates today (for the non-templating rules):

- No `CARGO_TARGET_DIR` setting: work with `./target`, regardless of if it's a symlink or directory, as is done today
- Using `CARGO_TARGET_DIR=/no-template`, ignore `./target` completely, as is done today
- Using `CARGO_TARGET_DIR=/tmp/{manifest-path-hash}`
  - `1.499`: ignore `./target` completely, as is done today, works in exactly `/tmp/{manifest-path-hash}`
  - `1.500`: works in `/tmp/<hash path>`
    - Forwarding:
      - `./target` is symlink: replace it, work with it
      - `./target` is not symlink: ignore it completely
    - No forwarding: ignore `./target` completely

Subsequent commands where the target dir (regardless of position) already exist would just work with it, as is done today: I can build with 1.76 and 1.77 in the same repo, it will just work in the same target dir and do whatever it wants with the artifacts and such.

But all of those rules (barring those that are already in action in released versions of `cargo`) shouldn't be set in stone IMO, I would prefer having users coming to us to tell us what they (didn't) like about them during the unstable period.

If enough people want to see them in the RFC then I can them though, since it will likely mean there is some consensus on them already